### PR TITLE
Introduce partlabel feature to find partition by their label

### PR DESCRIPTION
**Pull request by clementperon**
This has been tested under Zeus!

I encountered the issue when we change the disk to one of our machine to a NVMe instead of mSATA.

Acutally fstab and mender-client-resize-data-part use hardcoded path to the partition.

This can be an issue when you want to use the same image but on machine where the
disk enumeration is slightly different.

Example: one machine have a NVMe disk and the other a mSATA.
Or one machine have two hard drivers and you have installed mender on
the second one.

To avoid this we look for the correct partition based on the label device.

Note that the the grub.cfg and /data/mender/mender.conf should be correctly
configured by the user and should not rely on the image disk generated.

### CODE REVIEW

File: meta-mender-core/classes/mender-setup-image.inc [hunk 1]
```diff
@@ -32,9 +32,13 @@ mender_update_fstab_file() {
     local tmpBootPart="${MENDER_BOOT_PART}"
     local tmpDataPart="${MENDER_DATA_PART}"
     local partuuid="${@bb.utils.contains('MENDER_FEATURES', 'mender-partuuid', 'true', 'false', d)}"
+    local partlabel="${@bb.utils.contains('DISTRO_FEATURES', 'mender-partlabel', 'true', 'false', d)}"
     if [ ${partuuid} = "true" ]; then
         tmpBootPart="PARTUUID=${@mender_get_partuuid_from_device(d, '${MENDER_BOOT_PART}')}"
         tmpDataPart="PARTUUID=${@mender_get_partuuid_from_device(d, '${MENDER_DATA_PART}')}"
+    elif [ ${partlabel} = "true" ]; then
+        tmpBootPart="LABEL=boot"
+        tmpDataPart="LABEL=data"
```
Comment (lluiscampos) [inline meta-mender-core/classes/mender-setup-image.inc:41]: I suggest we set these in variables, defaulting to `boot` and `data` respectively. For example:
```suggestion
        tmpBootPart="LABEL=${MENDER_BOOT_PART_LABEL}"
        tmpDataPart="LABEL=${MENDER_DATA_PART_LABEL}"
```

And these being weak set in `mender-setup.bbclass` to "boot" and "data" with a comment explaining that are only used when `mender-partlabel` feature is enabled

File: meta-mender-core/recipes-mender/mender-client/files/mender-client-resize-data-part.sh.in [hunk 2]
```diff
@@ -4,8 +4,8 @@ set -e
 
 # All sizes are in blocks of 512 bytes
 
-MENDER_STORAGE_DEVICE=@MENDER_STORAGE_DEVICE@
 MENDER_DATA_PART=@MENDER_DATA_PART@
+MENDER_STORAGE_DEVICE=@MENDER_STORAGE_DEVICE@
```
Comment (lluiscampos) [inline meta-mender-core/recipes-mender/mender-client/files/mender-client-resize-data-part.sh.in:8]: Why was this needed?

File: meta-mender-core/recipes-mender/mender-client/mender-client.inc [hunk 3]
```diff
@@ -7,7 +7,7 @@ DEPENDS_class-native = ""
 
 PACKAGES_append = " mender-modules-gen"
 
-RDEPENDS_${PN}_append_mender-growfs-data_mender-systemd = " parted util-linux-fdisk"
+RDEPENDS_${PN}_append_mender-growfs-data_mender-systemd = " parted util-linux-fdisk util-linux-blkid"
```
Comment (lluiscampos) [inline meta-mender-core/recipes-mender/mender-client/mender-client.inc:10]: This dependency should be added only if `mender-partlabel` is enabled. Try with:
```suggestion
RDEPENDS_${PN}_append_mender-growfs-data_mender-systemd = " parted util-linux-fdisk"
RDEPENDS_${PN}_append_mender-growfs-data_mender-systemd_mender-partlabel = " util-linux-blkid"
```

File: meta-mender-core/recipes-mender/mender-client/mender-client.inc [hunk 4]
```diff
@@ -321,11 +321,20 @@ do_install_append_class-target_mender-image_mender-systemd() {
 }
 
 do_install_append_class-target_mender-growfs-data_mender-systemd() {
-    sed -i "s#@MENDER_STORAGE_DEVICE@#${MENDER_STORAGE_DEVICE}#g" \
-        ${WORKDIR}/mender-client-resize-data-part.sh.in
 
-    sed -i "s#@MENDER_DATA_PART@#${MENDER_DATA_PART}#g" \
-        ${WORKDIR}/mender-client-resize-data-part.sh.in
+    if ${@bb.utils.contains('DISTRO_FEATURES', 'mender-partlabel', 'true', 'false', d)}; then
```
Comment (lluiscampos) [inline meta-mender-core/recipes-mender/mender-client/mender-client.inc:325]: `MENDER_FEATURES` here too

File: meta-mender-core/recipes-mender/mender-client/mender-client.inc [hunk 5]
```diff
@@ -321,11 +321,20 @@ do_install_append_class-target_mender-image_mender-systemd() {
...
+        sed -i "s#@MENDER_DATA_PART@#\$(blkid -L data)#g" \
```
Comment (lluiscampos) [inline meta-mender-core/recipes-mender/mender-client/mender-client.inc:326]:
```suggestion
        sed -i "s#@MENDER_DATA_PART@#\$(blkid -L ${MENDER_DATA_PART_LABEL})#g" \
```

if my suggestion above goes in.

**Review by lluiscampos — CHANGES_REQUESTED** [review]
Hi @clementperon and thank you for your contribution.

It seems to me a good feature to have.

I made some comments on fixes and possible improvements. We also would need an automated acceptance test for this, but we can likely do it inside the Mender team (I would create a task for it).

Also, your commit needs to be signed and a changelog tag needs to be added, see [our guidelines](https://github.com/mendersoftware/mender/blob/master/CONTRIBUTING.md#providing-pull-requests).

File: meta-mender-core/classes/mender-bootimg.bbclass [hunk 6]
```diff
@@ -30,7 +30,7 @@ IMAGE_CMD_bootimg() {
             mkfs.${MENDER_BOOT_PART_FSTYPE_TO_GEN} \
                 $force_flag \
                 "${WORKDIR}/boot.${MENDER_BOOT_PART_FSTYPE_TO_GEN}" \
-                ${MENDER_BOOT_PART_FSOPTS} $label_flag "BOOT"
+                ${MENDER_BOOT_PART_FSOPTS} $label_flag "${MENDER_BOOT_PART_LABEL}"
```
Comment (clementperon) [inline meta-mender-core/classes/mender-bootimg.bbclass:33]: @lluiscampos Why do you think to use the $MENDER_BOOT_PART_LABEL in the bootimg class too?

File: meta-mender-core/recipes-mender/mender-client/files/mender-client-resize-data-part.sh.in (hunk 2 above)
Comment (clementperon) [reply meta-mender-core/recipes-mender/mender-client/files/mender-client-resize-data-part.sh.in:8]: The variable are interpreted top-down in shell.

And whe do:

```
MENDER_DATA_PART=$(blkid -L data)
MENDER_STORAGE_DEVICE="/dev/$(basename $(readlink -f "/sys/class/block/$(basename $MENDER_DATA_PART)/..))"
```

The readlink is a bit ugly we could do something like this instead:
```
MENDER_STORAGE_DEVICE="/dev/$(lsblk -no PKNAME $MENDER_DATA_PART)"
```
But this would required the lsblk utility

**Comment by clementperon** [comment]
@lluiscampos Thanks for your review.

Regarding the automated acceptance test, would be glad if this can be handle by mender team.

I have updated regarding your comment and also add a sanity check to forbid partlabel and partuuid at the same time.

File: meta-mender-core/recipes-mender/mender-client/files/mender-client-resize-data-part.sh.in (hunk 2 above)
Comment (lluiscampos) [reply meta-mender-core/recipes-mender/mender-client/files/mender-client-resize-data-part.sh.in:8]: Okey.

**Review by lluiscampos — COMMENTED** [review]
> @lluiscampos Thanks for your review.
> 

Sorry it took me so long to come back! After a week of vacation I had quite a bit to catch up this week...

> Regarding the automated acceptance test, would be glad if this can be handle by mender team.

Ok.

> 
> I have updated regarding your comment and also add a sanity check to forbid partlabel and partuuid at the same time.

It looks good to me. See one comment with regards to the default name of `MENDER_BOOT_PART_LABEL`. After fixing that I will run our CI on your changeset.

**Comment by lluiscampos** [comment]
@clementperon Please rebase also on top of latest master :+1:

File: meta-mender-core/classes/mender-bootimg.bbclass (hunk 6 above)
Comment (clementperon) [reply meta-mender-core/classes/mender-bootimg.bbclass:33]: blkid is case sensitive Changed to MENDER_BOOT_PART_LABEL ??= "${@bb.utils.contains('MENDER_BOOT_PART_FSTYPE', 'vfat', 'BOOT', 'boot', d)}"

**Comment by clementperon** [comment]
@lluiscampos rebased + changed to 
MENDER_BOOT_PART_LABEL ??= "${@bb.utils.contains('MENDER_BOOT_PART_FSTYPE', 'vfat', 'BOOT', 'boot', d)}"

**Comment by lluiscampos** [comment]
@mender-test-bot start pipeline

**Review by lluiscampos — APPROVED** [review]

**Comment by lluiscampos** [comment]
FYI @kacf, I created [MEN-5364](https://tracker.mender.io/browse/MEN-5364) epic to follow-up with testing.

Merging now, thank you @clementperon for your contribution :rocket:
